File: 2017/day25.py
# @formatter:off
from os.path import isfile, join as path_join
from sys import path as sys_path
sys_path.insert(1, path_join(sys_path[0], '..'))
file_name = path_join('input', 'day25.txt')
def to_list(mf: object = int, sep='\n'): return [mf(x) for x in open(file_name).read().split(sep) if x]
def to_gen(mf: object = int, sep='\n'): return (mf(x) for x in open(file_name).read().split(sep) if x)
def p1(*args): ans = part1(*args); print(ans); return ans
def p2(*args): ans = part2(*args); print(ans); return ans

if not isfile(file_name):
	from aoc import get_input_file
	get_input_file(session_path=['..', '.env'])
# @formatter:on

from utils import parsefile

parts = parsefile(file_name, [[str, "\n"], "\n\n"])

begin_state = parts[0][0].split(" ")[-1][:-1]
steps = int(parts[0][1].split(" ")[5])

states = {}

for p in parts[1:]:
	id = p[0].split(" ")[-1][:-1]

	zero_cond = []
	zero_cond.append(int(p[2].split(" ")[-1][:-1]))  # write value
	zero_cond.append(1 if p[3].split(" ")[-1][:-1] == "right" else -1)  # move dir
	zero_cond.append(p[4].split(" ")[-1][:-1])  # next state

	one_cond = []
	one_cond.append(int(p[6].split(" ")[-1][:-1]))  # write value
	one_cond.append(1 if p[7].split(" ")[-1][:-1] == "right" else -1)  # move dir
	one_cond.append(p[8].split(" ")[-1][:-1])  # next state

	states[id] = (zero_cond, one_cond)


def part1():
	tape = {}
	pos = 0
	state = begin_state

	for i in range(steps):
		if pos in tape:
			value = tape[pos]
		else:
			value = 0

		cs = states[state]

		# write value
		tape[pos] = cs[value][0]

		# move the pos
		pos += cs[value][1]

		# change the state
		state = cs[value][2]

	c = 0

	for p in tape:
		if tape[p] == 1:
			c += 1

	return c


def part2():
	pass


p1()
p2()

File: cpp/2017/day25.cpp
#include "../aocHelper.h"

class Day25 : public BaseDay
{
public:
	Day25() : BaseDay("25") {}

	result_type solve() override
	{
		long long part1 = 0, part2 = 0;

		struct data
		{
			int value;
			int dir;
			int next;
		};

		using state = array<data, 2>;

		//unordered_map<char, state> states;
		array<state, 6> states;

		input += 15; // skip "Begin in state "
		int start_state = *input - 'A';
		input++;
		input += 2; // skip ".\n"
		input += 36; // skip "Perform a diagnostic checksum after "
		int steps = numericParse<int>(input);
		input += 8; // skip " steps.\n"

		// parse input
		while (*input != '\0')
		{
			state st;

			input++; // skip '\n'
			input += 9; // skip "In state "
			char c = *input;
			input++;
			input += 2; // skip ":\n"
			input += 29; // skip "  If the current value is 0:\n"
			input += 22; // skip "    - Write the value "
			st[0].value = numericParse<int>(input);
			input += 2; // skip ".\n"
			input += 27; // skip "    - Move one slot to the "
			if (*input == 'r')
			{
				st[0].dir = 1;
				input += 5; // skip "right"
			}
			else
			{
				st[0].dir = -1;
				input += 4; // skip "left"
			}
			input += 2; // skip ".\n"
			input += 26; // skip "    - Continue with state "
			st[0].next = *input - 'A';
			input++;
			input += 2; // skip ".\n"

			input += 29; // skip "  If the current value is 1:\n"
			input += 22; // skip "    - Write the value "
			st[1].value = numericParse<int>(input);
			input += 2; // skip ".\n"
			input += 27; // skip "    - Move one slot to the "
			if (*input == 'r')
			{
				st[1].dir = 1;
				input += 5; // skip "right"
			}
			else
			{
				st[1].dir = -1;
				input += 4; // skip "left"
			}
			input += 2; // skip ".\n"
			input += 26; // skip "    - Continue with state "
			st[1].next = *input - 'A';
			input++;
			input += 2; // skip ".\n"

			//states.insert({ c, st });
			states[c - 'A'] = st;
		}

		// part 1
		unordered_map<int, int> tape;
		int pos = 0;
		int curr_state = start_state;

		for (int i = 0; i < steps; i++)
		{
			int value = 0;
			if (tape.contains(pos))
			{
				value = tape[pos];
			}

			state& st = states[curr_state];

			// write value
			tape[pos] = st[value].value;

			// move the pos
			pos += st[value].dir;

			// change the state
			curr_state = st[value].next;
		}

		for (auto& p : tape)
		{
			if (p.second == 1)
			{
				part1++;
			}
		}

		return { part1, part2 };
	}
};
